src/main/java/com/aws/greengrass/util/platforms/windows/WindowsRunasProcess.java

```diff
@@ -0,0 +1,450 @@
+        startupInfo.dwFlags = WinBase.STARTF_USESTDHANDLES;
+        startupInfo.hStdInput = inPipeReadHandle.getValue();
+        startupInfo.hStdOutput = outPipeWriteHandle.getValue();
+        startupInfo.hStdError = errPipeWriteHandle.getValue();
+        startupInfo.write();
+        return startupInfo;
+    }
+
+    @SuppressWarnings("PMD.AvoidFileStream")
+    private void redirectStreams() throws ProcessCreationException {
+        FileDescriptor stdinFd = new FileDescriptor();
+        FileDescriptor stdoutFd = new FileDescriptor();
+        FileDescriptor stderrFd = new FileDescriptor();
+        writefd(stdinFd, inPipeWriteHandle.getValue());
+        writefd(stdoutFd, outPipeReadHandle.getValue());
+        writefd(stderrFd, errPipeReadHandle.getValue());
+        stdin = new FileOutputStream(stdinFd);
+        stdout = new FileInputStream(stdoutFd);
+        stderr = new FileInputStream(stderrFd);
+    }
+
+    private void closeHandles() {
+        if (outPipeWriteHandle != null) {
+            try {
+                Kernel32Util.closeHandleRefs(outPipeWriteHandle, outPipeReadHandle, errPipeWriteHandle,
+                        errPipeReadHandle, inPipeReadHandle, inPipeWriteHandle);
+            } catch (Win32Exception e) {
+                // Nothing we can do. We made best effort to close resources
+            }
+        }
+        if (procInfo != null) {
+            try {
+                Kernel32Util.closeHandles(procInfo.hProcess, procInfo.hThread);
+            } catch (Win32Exception e) {
+                // Nothing we can do. We made best effort to close resources
+            }
+        }
+    }
+
+    @Override
+    public OutputStream getOutputStream() {
+        return stdin;
+    }
+
+    @Override
+    public InputStream getInputStream() {
+        return stdout;
+    }
+
+    @Override
+    public InputStream getErrorStream() {
+        return stderr;
+    }
+
+    @Override
+    public synchronized int waitFor() throws InterruptedException {
+        if (procInfo == null) {
+            throw new RuntimeException("process was not created");
+        }
+        if (exited) {
+            return exitCode;
+        }
+        if (WinBase.WAIT_FAILED == Kernel32.INSTANCE.WaitForSingleObject(procInfo.hProcess, Kernel32.INFINITE)) {
+            throw lastErrorRuntimeException();
+        }
+        return exitValue();
+    }
+
+    @Override
+    public synchronized int exitValue() {
+        if (procInfo == null) {
+            throw new RuntimeException("process was not created");
+        }
+        if (exited) {
+            return exitCode;
+        }
+        IntByReference exitCodeRef = new IntByReference();
+        if (!Kernel32.INSTANCE.GetExitCodeProcess(procInfo.hProcess, exitCodeRef)) {
+            throw lastErrorRuntimeException();
+        }
+        if (WinBase.STILL_ACTIVE == exitCodeRef.getValue()) {
+            throw new IllegalThreadStateException("process hasn't exited");
```

Comment:
why `IllegalThreadStateException`?

Reply:
That is what javadoc says the API should do if it hasn't exited